pkg/operator/cluster/creds_controller.go

```diff
@@ -0,0 +1,176 @@
+	// in seconds
+	controllerResyncPeriod time.Duration = 120
+	// controllerRetry is number of retries to process an event
+	controllerRetry int = 5
+)
+
+func NewCredsController(ns string, context *clusterd.Context) *PVCController {
+	return &PVCController{
+		clusterNamespace: ns,
+		clusterContext:   context,
+		queue:            workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+	}
+}
+
+// PVCController implements Kube controller that monitors rook PVC events.
+// The controller ensures all kube namespaces have a set of credentials
+// to manage lifeCycle of ceph primitives.
+type PVCController struct {
+	queue            workqueue.RateLimitingInterface
+	indexer          cache.Indexer
+	informer         cache.Controller
+	clusterContext   *clusterd.Context
+	clusterNamespace string
+}
+
+func (c *PVCController) pvcHandler(key interface{}) error {
+	obj, _, err := c.indexer.GetByKey(key.(string))
+	if err != nil {
+		return err
+	}
+
+	pvc := obj.(*v1.PersistentVolumeClaim)
+
+	if !strings.Contains(*pvc.Spec.StorageClassName, "rook") {
+		return nil
+	}
+
+	cephUser := &cephUser{c.clusterContext, c.clusterNamespace}
+	if err := cephUser.create(); err != nil {
+		return err
+	}
+
+	if err = cephUser.setKubeSecret(pvc.ObjectMeta.Namespace); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *PVCController) run(stopCh chan struct{}) {
+	eventList := cache.NewListWatchFromClient(
+		c.clusterContext.Clientset.Core().RESTClient(),
+		"persistentvolumeclaims",
+		v1.NamespaceAll,
+		fields.Everything(),
+	)
+
+	c.indexer, c.informer = cache.NewIndexerInformer(
+		eventList,
+		&v1.PersistentVolumeClaim{},
+		controllerResyncPeriod*time.Second,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.onAdd,
+			UpdateFunc: c.onUpdate,
+			DeleteFunc: c.onDelete,
+		},
+		cache.Indexers{},
+	)
+
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+	logger.Info("Starting PVC controller (managing ceph creds).")
+
+	go c.informer.Run(stopCh)
+
+	// Wait for all involved caches to be synced, before processing items from the queue is started
+	if !cache.WaitForCacheSync(stopCh, c.informer.HasSynced) {
+		runtime.HandleError(fmt.Errorf("Timed out waiting for caches to sync"))
+		return
+	}
+
+	go wait.Until(c.runWorker, time.Second, stopCh)
+
+	<-stopCh
+	logger.Info("Stopping PVC (ceph creds) controller.")
+}
+
+func (c *PVCController) runWorker() {
+	for c.processNextItem() {
+	}
+}
+
+func (c *PVCController) processNextItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+
+	defer c.queue.Done(key)
+
+	err := c.pvcHandler(key)
+	c.handleErr(err, key)
+	return true
+}
+
+func (c *PVCController) onAdd(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err == nil {
+		c.queue.Add(key)
+	}
+}
+
+func (c *PVCController) onUpdate(obj, oldObj interface{}) {
+	logger.Debug("Not implemented. Checking Ceph Creds on pvc UPDATE events.")
```

Comment:
Left it _unimplemented_ for now since there is significant amount of update events possibly from rook operator or other rook component. Planned to put more though and investigation into it.